### PR TITLE
Initial pokemon upload working.

### PR DESCRIPTION
**Pull request by YaManicKill**
I should probably add some more tests, and there's a few UI things that need done, but we can probably merge as is just now and I'll do another PR for that, if you are happy.

### CODE REVIEW

File: api/controllers/PokemonController.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ module.exports = _.mapValues({
     let box;
     if (params.box) {
       box = await Box.findOne({id: params.box});
+      console.log(box);
```
Comment (not-an-aardvark) [inline api/controllers/PokemonController.js]: We can probably get rid of this line

File: api/controllers/HomeController.js [hunk 2]
```diff
@@ -8,7 +8,8 @@
 module.exports = _.mapValues({
   async index (req, res) {
     const boxes = await Box.find({owner: req.user.name});
-    res.view('home/view', {boxes});
+    const prefs = await UserPreferences.find({user: req.user.name});
```
Comment (not-an-aardvark) [inline api/controllers/HomeController.js]: If you replace this with `UserPreferences.findOne`, then the client can do `this.prefs.defaultPokemonVisibility` instead of `this.prefs[0].defaultPokemonVisibility`.

**Comment by not-an-aardvark** [comment]
A few UI things I noticed, which probably don't have to be fixed right away:

* The "Choose a pk6 or pkx file." error message pops up as soon as I click "Choose file" -- I'm not sure if that's intended.

* The box titles seem to have scrollbars now, which [looks a bit weird](https://i.gyazo.com/c7ab419bc46d1a88f85a62f6465e5e60.png) in my opinion. We might want to make the font size so that at least two lines of the title are visible.

* If no box is specified, the server creates a new box for the user, but there's no way to see the box without refreshing the page. (This might be partially a serverside issue.) Also, we might want to clarify to the user what happens if they don't pick a box.

* After uploading, there's no indication that the upload was successful. It might be a good idea to redirect to the newly-uploaded pokemon's page (although we don't have a view for that yet).

* If I have the upload menu up and I have the box dropdown open, and I press "Esc", the entire upload menu closes and I lose whatever file I had selected. It would be better if it only the box dropdown closed.

* The upload menu [doesn't quite fit the screen](https://i.gyazo.com/cb101fcfe88eb228402d7472c95b3733.png).

File: api/controllers/HomeController.js (hunk 2 above)
Comment (not-an-aardvark) [inline api/controllers/HomeController.js]: Re. rendering this stuff serverside -- I think we shouldn't touch the serverside code at all for the purposes of creating the client, and instead use the API to get the preferences (via the `/preferences` endpoint). 

Advantages of not touching the serverside code:

* Our client and server end up less coupled.
* We already have a JSON API for this info, so we don't need to maintain two separate methods to access it.
* There are already thorough tests in place for the JSON API, particularly for security concerns (e.g. not disclosing the existence of a user's unlisted boxes). If the server is also sending database information via the HTML site, then we have twice as many things to test, and we're much more likely to mess it up somewhere and leak private info, like what happened [here](/pokemontrades/flairhq/pull/340/files#diff-c027fe067a643ce60d09a3aff3f4cde9L14) with FlairHQ. The safest way to keep the info secure is to have as little attack surface as possible.

tl;dr: We already have a RESTful API, let's use it.

File: client/add/pokemon.ctrl.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    this.file.upload = Upload.upload({
+      url: '/uploadpk6',
+      data: {visibility: this.visibility, pk6: this.file, box: this.box},
+    }).then(function (response) {
+      $mdDialog.hide(response.data);
+    }, function (response) {
+      console.log(response.status + ': ' + response.data);
+    }, function (evt) {
+      // Math.min is to fix IE which reports 200% sometimes
+      console.log(Math.min(100, parseInt(100.0 * evt.loaded / evt.total)));
```
Comment (not-an-aardvark) [inline client/add/pokemon.ctrl.js]: I like the idea of a progress meter, but these uploads are all around 250 bytes -- I'm not sure it's worth displaying it since it'll be at 100 almost immediately anyway.

Obviously, if we do decide to display it then we should probably do something other than printing it to the console.

File: api/controllers/PokemonController.js (hunk 1 above)
Comment (YaManicKill) [inline api/controllers/PokemonController.js]: Oh yeah, it was for testing my box uploading was working (because we don't have any way to view the box yet).

File: api/controllers/HomeController.js (hunk 2 above)
Comment (YaManicKill) [inline api/controllers/HomeController.js]: So are you suggesting we go back to how flairhq did things originally (which we changed to this way of doing things because it was quicker)?

File: client/add/pokemon.ctrl.js (hunk 3 above)
Comment (YaManicKill) [inline client/add/pokemon.ctrl.js]: Yeah, I don't know why I have that there, it's probably just copied code from the example on the webpage. I'll remove it.

**Comment by YaManicKill** [comment]
> The "Choose a pk6 or pkx file." error message pops up as soon as I click "Choose file" -- I'm not sure if that's intended.

Yeah, that's always going to be a problem, I think. The problem is that it shows the error if you have touched the element and not got something legitimate yet. I don't think there's a real way around it.

> The box titles seem to have scrollbars now, which looks a bit weird in my opinion. We might want to make the font size so that at least two lines of the title are visible.

Hmmm, how did that happen? I thought I would truncating them.

> If no box is specified, the server creates a new box for the user, but there's no way to see the box without refreshing the page. (This might be partially a serverside issue.) Also, we might want to clarify to the user what happens if they don't pick a box.

Yeah, I meant to chat to you about this. I don't think we should create a new box. It seems a bit of an odd way. We can probably have a default box they have to set? Or like thought of before, we can have it default to a "transport" box if they aren't in a box?

> After uploading, there's no indication that the upload was successful. It might be a good idea to redirect to the newly-uploaded pokemon's page (although we don't have a view for that yet).

I'm leaving this till box view is done.

> If I have the upload menu up and I have the box dropdown open, and I press "Esc", the entire upload menu closes and I lose whatever file I had selected. It would be better if it only the box dropdown closed.

Oh. Hmmmm. I'll look into that. I'm probably going to move to a modal for larger screens anyway, so that won't be an issue there/

> The upload menu doesn't quite fit the screen.

Hmmm, I wasn't getting that. Weird. Again, won't be an issue if I use a modal instead on desktop (which is much nicer for desktop).

**Comment by YaManicKill** [comment]
I'll do the few things I responded to, and fix the tests/write some more. And then we can merge in.

File: api/controllers/HomeController.js (hunk 2 above)
Comment (not-an-aardvark) [inline api/controllers/HomeController.js]: Yeah, on reconsideration I think this might the better way, provided that we don't block the page load while waiting for all the requests to come in. (For example, we can render the page without knowing the user's default upload preferences.)

I suppose another way to do it would be to have batch endpoints for all the resources one would need for rendering (to keep the number of requests down) and then test those as well to make sure they don't leak info. I feel like that still duplicates a lot of functionality though, and it increases the attack surface.

**Comment by not-an-aardvark** [comment]
> Yeah, I meant to chat to you about this. I don't think we should create a new box. It seems a bit of an odd way. We can probably have a default box they have to set? Or like thought of before, we can have it default to a "transport" box if they aren't in a box?

Sure, I'd wouldn't mind changing that.

**Comment by not-an-aardvark** [comment]
> > The "Choose a pk6 or pkx file." error message pops up as soon as I click "Choose file" -- I'm not sure if that's intended.

> Yeah, that's always going to be a problem, I think. The problem is that it shows the error if you have touched the element and not got something legitimate yet. I don't think there's a real way around it.

Does there need to be an error message at all? The file upload window only allows pk6 files anyway.

**Comment by YaManicKill** [comment]
> Does there need to be an error message at all? The file upload window only allows pk6 files anyway.

Good point, I guess. I'll remove it for that one for now.

> Sure, I'd wouldn't mind changing that.

Which one do you think is best? I'd be happy with either. Transport box keeps the Pokemon Bank methodology which people know. Default box is perhaps nicer? We could even have a "transport box" created by default which is people's default default box (I think) which they can delete and change to a different box.

**Comment by not-an-aardvark** [comment]
Another option (which may or may not be the best one) would be to simply not allow users to upload without selecting a box, and then have one (non-special) box automatically get created when a user creates an account. A transport box could also be a good idea, but people might not like having a box that they're unable to delete.

**Comment by YaManicKill** [comment]
> simply not allow users to upload without selecting a box

Could do.

> people might not like having a box that they're unable to delete.

I actually suggested them being able to delete it if they set a different box as default.

**Comment by not-an-aardvark** [comment]
> I actually suggested them being able to delete it if they set a different box as default.

Oops, I missed that.

I think I like the default box idea, but what happens if someone tries to delete their default box? Does the "default" status move to another box?

**Comment by YaManicKill** [comment]
> I think I like the default box idea, but what happens if someone tries to delete their default box? Does the "default" status move to another box?

2 Options:

1. don't let them delete their default box  
2. Ask them what box they want to be their default (and if they don't, choose the first)

I'm not sure what I prefer. I think we should take this to a ticket. Outside of the scope of this maybe?

**Comment by not-an-aardvark** [comment]
Sure let's do that, it's not really related to this PR anyway.

File: api/controllers/HomeController.js (hunk 2 above)
Comment (YaManicKill) [inline api/controllers/HomeController.js]: Let's keep it as it is just now, and once we have alpha up and running we can see if having all this is causing issues. I doubt it will, as the boxes is the only big one, and we need that to show anything anyway.

Comment (not-an-aardvark) [inline api/controllers/HomeController.js]: Ok, we can keep this one as it is, but let's try to use the existing API as much as possible for future stuff (e.g. viewing other users' boxes, since there are actually security concerns for those).
